README.md
<p align="center">
	<a href="https://scrollrevealjs.org" title="Visit ScrollReveal home page">
		<img src="https://scrollrevealjs.org/img/logomark.svg" alt="ScrollReveal" width="120">
	</a>
</p>
<br>
<p align="center">
	<a href="https://scrollrevealjs.org" title="Visit ScrollReveal home page">
		<img width="200" src="https://scrollrevealjs.org/img/scrollreveal-logotype-dark.svg" alt="ScrollReveal">
	</a>
</p>
<p align="center">Animate elements as they scroll into view.</p>

<p align="center">
	<a href="https://travis-ci.org/jlmakes/scrollreveal">
		<img src="https://img.shields.io/travis/jlmakes/scrollreveal.svg" alt="Build status">
	</a>
	<a href="https://www.npmjs.com/package/scrollreveal">
		<img src="https://img.shields.io/npm/dm/scrollreveal.svg" alt="Monthly downloads">
	</a>
	<a href="https://www.npmjs.com/package/scrollreveal">
		<img src="https://img.shields.io/npm/v/scrollreveal.svg" alt="Version">
	</a>
	<img src="https://img.shields.io/badge/min+gzip-5.7_kB-blue.svg" alt="5.7 kB min+gzip">
	<a href="https://opensource.org/licenses/GPL-3.0">
		<img src="https://img.shields.io/badge/license-GPLv3-blue.svg" alt="GPLv3 License">
	</a>
</p>

<p align="center">
	<a href="https://saucelabs.com/u/scrollreveal">
		<img src="https://saucelabs.com/browser-matrix/scrollreveal.svg?y=1" alt="Browser compatibility matrix" width="100%">
	</a>
</p>

<br>

<br>

# Introduction

ScrollReveal is a JavaScript library for easily animating elements as they enter/leave the viewport. It was designed to be robust and flexible, but hopefully you’ll be surprised below at how easy it is to pick up.

<br>

# Installation

## Browser

A simple and fast way to get started is to include this script on your page:

```html
<script src="https://unpkg.com/scrollreveal"></script>
```

This will create the global variable `ScrollReveal`

> Be careful using this method in production. Without specifying a fixed version number, Unpkg may delay your page load while it resolves the latest version. Learn more at [unpkg.com](https://unpkg.com)

## Module

```bash
npm install scrollreveal
```

#### CommonJS

```js
const ScrollReveal = require('scrollreveal')
```

#### ES2015

```js
import ScrollReveal from 'scrollreveal'
```

<br>

# Usage

Installation provides us with the constructor function [`ScrollReveal()`](https://scrollrevealjs.org/api/constructor.html). Calling this function returns the ScrollReveal instance, the “brain” behind the magic.

> ScrollReveal employs the singleton pattern; no matter how many times the constructor is called, it will always return the same instance. This means we can call it anywhere, worry-free.

There’s a lot we can do with this instance, but most of the time we’ll be using the [`reveal()`](https://scrollrevealjs.org/api/reveal.html) method to create animation. Fundamentally, this is how to use ScrollReveal:

```html
<h1 class="headline">
	Widget Inc.
</h1>
```

```js
ScrollReveal().reveal('.headline')
```

**🔎 See this demo live on [JSBin](http://jsbin.com/jufohaxonu/edit?html,output)**

<br>

---

### The full documentation can be found at [https://scrollrevealjs.org](https://scrollrevealjs.org)

> If you’re using an older version of ScrollReveal, you can find legacy documentation in the [wiki](https://github.com/jlmakes/scrollreveal/wiki)

---

<br>

<a href="https://scrollrevealjs.org/pricing/" title="Visit ScrollReveal pricing page">
	<img align="right" height="300" src="https://scrollrevealjs.org/img/license.svg" alt="Commercial License Badge">
</a>

<br>

# License

**For commercial sites, themes, projects, and applications, keep your source code private/proprietary by purchasing a [Commercial License](https://scrollrevealjs.org/pricing/).**

Licensed under the GNU General Public License 3.0 for compatible open source projects and non-commercial use.

<br>

Copyright 2021 Fisssion LLC


CHANGELOG.md
# Change Log

## [4.0.9] - 2021-03-04

### Fixed

- Styles applied using CSSOM don't drop `:` characters.

## [4.0.8] - 2021-03-02

### Fixed

- Avoid Content Security Policy (CSP) violations. [@lambdacasserole](https://github.com/lambdacasserole) [#431](https://github.com/jlmakes/scrollreveal/pull/431)

## [4.0.7] - 2020-07-15

### Fixed

- Ensure element geometry exists. [#437](https://github.com/jlmakes/scrollreveal/issues/437)

## [4.0.6] - 2020-03-15

### Fixed

- Default transition values of `none` are now correctly ignored. [#231](https://github.com/jlmakes/scrollreveal/issues/231)

### Fixed

## [4.0.5] - 2018-10-20

### Fixed

- Calling `reveal()` on the same `target` breaking animation. [#468](https://github.com/jlmakes/scrollreveal/issues/468)

## [4.0.4] - 2018-09-22

### Fixed

- Malformed `package.json`

## [4.0.3] - 2018-09-21

### Fixed

– `options.cleanup` is now correctly set to `false` by default. [#457](https://github.com/jlmakes/scrollreveal/issues/457)

## [4.0.2] - 2018-09-11

### Fixed

- Null property assignment regression in mount function. [#456](https://github.com/jlmakes/scrollreveal/issues/456)

## [4.0.1] - 2018-09-09

### Fixed

- Noop instances were not correctly unmounting from the DOM. [#455](https://github.com/jlmakes/scrollreveal/issues/455)
- Readme links to pricing page no longer 404.

## [4.0.0] - 2018-08-06

### Added

- ScrollReveal can be enabled/disabled on desktops using `options.desktop`.
- The class `sr` is added to `<html>` during instantiation when supported. [#294](https://github.com/jlmakes/scrollreveal/issues/294)
- `height: 100%` is added to `<body>` during instantiation when supported. [#298](https://github.com/jlmakes/scrollreveal/issues/298)
- Unused containers are removed from the store, and their event listeners destroyed.
- ScrollReveal skips generating opacity styles when `options.opacity` is set to `null`.
- ScrollReveal retains element CSS transformations. [#251](https://github.comjlmakes/scrollreveal/issues/251)
- New `options.cleanup` toggles whether generated styles are removed upon reveal completion (when `options.reset` is `false`). [#292](https://github.comjlmakes/scrollreveal/issues/292)
- ScrollReveal tracks scroll direction as container store data. [#384](https://github.com/jlmakes/scrollreveal/issues/384)
- New `clean()` method removes specific generated styles and event listeners. [#227](https://github.com/jlmakes/scrollreveal/issues/227)
- New `destroy()` method removes all generated styles and event listeners. [#227](https://github.com/jlmakes/scrollreveal/issues/227)
- New `debug` static property toggles error messages in console. [#351](https://github.com/jlmakes/scrollreveal/issues/351)
- Instance methods now accept native arrays of HTML elements.

### Changed

- **Breaking:** The `reveal()` method no longer accepts an `interval` parameter. Instead, sequence intervals are now defined with `options.interval`.
- **Breaking:** The instance method `isSupported()` is now static.
- **Breaking:** `options.distance` supports only `em` `px` and `%` values.
- **Breaking:** ScrollReveal methods are no longer chainable.
- **Breaking:** ScrollReveal requires a commercial license, unless for [GPL-3.0](https://opensource.org/licenses/GPL-3.0) compatible open source projects.
- Elements in a reveal sequence are no longer grouped, and reveal progressively when visible.
- ScrollReveal uses a single `matrix3d()` property, with the correct prefix and only when necessary. [#292](https://github.com/jlmakes/scrollreveal/issues/292)
- ScrollReveal returns a non-operational instance when instantiated in unsupported browsers.
- ScrollReveal `version` is now a read-only instance property.
- ScrollReveal methods are now bound read-only instance properties.
- `options.viewFactor` clamps values outside of `0.0` to `1.0`.
- ScrollReveal constructor now returns a singleton.

### Fixed

- The `requestAnimationFrame` polyfill now reliably throttles callback invocations.

## [3.3.6] - 2017-06-23

### Fixed

- Element visibility now checks left and right boundaries correctly. [#352](https://github.com/jlmakes/scrollreveal/issues/352)
- Library version instance property is again accurate.

## [3.3.5] - 2017-04-05

### Fixed

- Patched to ensure version 3 is the default NPM package.

## [3.3.4] - 2017-02-18


scrollReveal.js
/*
           _____                 ________                       __
          / ___/______________  / / / __ \___ _   _____  ____ _/ /
          \__ \/ ___/ ___/ __ \/ / / /_/ / _ \ | / / _ \/ __ `/ /
         ___/ / /__/ /  / /_/ / / / _, _/  __/ |/ /  __/ /_/ / /
        /____/\___/_/   \____/_/_/_/ |_|\___/|___/\___/\__,_/_/   v3.0.0

________________________________________________________________________________

      Copyright 2015 Julian Lloyd (@jlmakes) Open source under MIT license
______________________________________________________________________________*/

window.ScrollReveal = (function( window ){

  'use strict';

  var _requestAnimFrame
    , _extend
    , _extendClone
    , _handler
    , _isDOMElement
    , _updateElemStore
    , self;

  function ScrollReveal( config ){

    self = this;

    _extend( self.defaults, config );

    if ( self.isMobile() && !self.defaults.mobile || !self.isSupported() ){

      // Remove any data-sr attributes found in the DOM and abort

      self.cleanDOM();
      console.warn('ScrollReveal instantiation aborted.');
      return null;
    }

    self.store = {
      elements:  {}, // Element store to manage styles
      viewports: []  // Viewport store to manage event binding
    };

    self.serial      = 1;     // Primary key for element store
    self.blocked     = false; // Flag used to throttle scroll events
    self.initialized = false; // Flag used for 'onload' delay styles
  }



  ScrollReveal.prototype.defaults = {

    enter:       'bottom', // top, right, bottom, left
    move:        '0px',
    over:        '0.75s',
    wait:        '0s',
    easing:      'ease',

    scale:       { direction: 'up', power: '10%' },
    rotate:      { x: 0, y: 0, z: 0 },

    opacity:     0,    // The starting opacity for reveal animations.
    mobile:      true, // Controls whether or not reveals occur on mobile

    //           Controls whether or not reveal animations play each time the
    //           element enters its viewport, or just once
    reset:       false,

    //           Expects a reference to a DOM node (the <html> node by default)
    //           which is used as the context when checking element visibility
    viewport:    window.document.documentElement,

    //           'always' — Delay every time an animation resets
    //           'onload' - Delay only for animations triggered by first load
    //           'once'   — Delay only the first time an animation reveals
    delay:       'once',

    //           vFactor changes when an element is considered in the viewport
    //           The default value of 0.60 means 60% of an element must be
    //           visible for its reveal animation to trigger
    vFactor:     0.60,

    //           Callbacks:
    afterReveal: function( domEl ){},
    afterReset:  function( domEl ){}
  };



  ScrollReveal.prototype.init = function(){

    var viewport;

    // Register any inline ScrollReveal instructions

    self.reveal.call( self.init, '[data-sr]' );

    // Go through the viewport store, and bind event listeners

    for ( var i = 0; i < self.store.viewports.length; i++ ){

      viewport = self.store.viewports[ i ];

      if ( viewport === window.document.documentElement ){

        window.addEventListener( 'scroll', _handler, true );

      } else {

        viewport.addEventListener( 'scroll', _handler, true );
      }
    }

    if ( !self.initialized ){
      window.addEventListener( 'resize', _handler, true );
      self.initialized = true;
    }

    self.animate();

    return self;
  };



  ScrollReveal.prototype.reveal = function( selector, config ){

    var elem, elems, viewport;

    if ( selector == '[data-sr]' && this != self.init ){
      return console.warn("Invalid selector [data-sr] passed to reveal method.");
    }

    if ( config && config.viewport ){

      viewport = config.viewport;

    } else {

      viewport = self.defaults.viewport;
    }

    elems =
      Array
        .prototype
        .slice
        .call( viewport.querySelectorAll( selector ) );

    // If no elements are found, display warning message in console and exit

    if ( elems.length == 0 && selector != '[data-sr]' ){
      console.warn( selector + " inside " + config.viewport + " returned 0 elements." );
      return
    }

    for ( var i = 0; i < elems.length; i++ ){

      var elem = {}
        , id   = elems[ i ].getAttribute('data-sr-id');

      if ( id ){

        // If we find an element, populate our element
        // with the object at that key in the data store

        elem = self.store.elements[ id ];

      } else {

        // Otherwise, create a new element.

        elem.domEl    = elems[ i ];
        elem.id       = self.serial++;
        elem.seen     = false;
        elem.revealed = false;

        elem.domEl.setAttribute( 'data-sr-id', elem.id );
      }

      // Now that we have an element, let’s update its config and styles

      if ( this == self.init ){

        elem.config = self.configFactory( elem.domEl.getAttribute('data-sr'), elem.config );

      } else {

        elem.config = self.configFactory( config, elem.config );
      }

      elem.styles = self.styleFactory( elem );

      elem.domEl.setAttribute( 'style',
          elem.styles.inline
        + elem.styles.initial
      );

      _updateElemStore( elem );
    }

    return self;
  };



  ScrollReveal.prototype.animate = function(){

    var elem
      , key
      , visible;

    // Begin element store digest

    for ( key in self.store.elements ){
      if ( self.store.elements.hasOwnProperty( key ) ){

        elem    = self.store.elements[ key ];
        visible = self.isElemVisible( elem );

        if ( visible && !elem.revealed ){

          if ( elem.config.delay === 'always'
          || ( elem.config.delay === 'onload' && !self.initialized )
          || ( elem.config.delay === 'once'   && !elem.seen ) ){

            // Use animation with delay

            elem.domEl.setAttribute( 'style',
                elem.styles.inline
              + elem.styles.target
              + elem.styles.transition
            );

          } else {

            // Use animation without delay

            elem.domEl.setAttribute( 'style',
                elem.styles.inline
              + elem.styles.target
              + elem.styles.reset
            );
          }

          elem.seen = true;
          queueCallback( 'reveal', elem );
        }

        else if ( !visible && elem.config.reset && elem.revealed ){

          elem.domEl.setAttribute( 'style',
              elem.styles.inline
            + elem.styles.initial
            + elem.styles.reset
          );

          queueCallback( 'reset', elem );
        }
      }
    }

    // Unblock the event handler and register the first initialization

    self.blocked = false;

    function queueCallback( type, elem ){

      var elapsed  = 0
        , duration = elem.styles.duration[ type ]
        , callback = "after" + type.charAt(0).toUpperCase() + type.slice(1);

      // Check if element already has a running timer, and capture the elapsed
      // time so we can offset the animation duration

      if ( elem.timer ){
        elapsed = Math.abs( elem.timer.started - new Date() );
        window.clearTimeout( elem.timer.clock );
      }

      elem.timer = { started: new Date() };

      elem.timer.clock = window.setTimeout(function(){

        elem.config[ callback ]( elem.domEl );
        elem.timer = null;

      }, duration - elapsed );

      return ( type == "reveal" ) ? elem.revealed = true : elem.revealed = false;
    }
  };



  ScrollReveal.prototype.configFactory = function( config, context ){

    // The default configuration is the default context, but in instances
    // where we call sr.reveal() more than once on the same element set
    // (perhaps to re-configure or override), we need to set the context
    // to the element’s existing styles

    var words
      , parsed = {};

    // Confirm our context

    if ( context == null ){
      context = self.defaults;
    }

    else if ( context && typeof context !== 'object' || context.constructor != Object ){
      context = self.defaults;
    }

    // Confirm configuration

    if ( config == null ) {
      config = context;
    }

    else if ( typeof config === 'object' && config.constructor == Object ){
      config = _extendClone( context, config );
    }

    else if ( typeof config === 'string' ){

      words = config.split( /[, ]+/ );
      words.forEach(function( word, i ){
        switch ( word ){

          case 'enter':

            parsed.enter = words[ i + 1 ];
            break;

          case 'wait':

            parsed.wait = words[ i + 1 ];
            break;

          case 'move':

            parsed.move = words[ i + 1 ];
            break;

          case 'ease':

            parsed.move = words[ i + 1 ];
            parsed.ease = 'ease';
            break;

          case 'ease-in':

            if ( words[ i + 1 ] == 'up'
              || words[ i + 1 ] == 'down' ){

              parsed.scale.direction = words[ i + 1 ];
              parsed.scale.power     = words[ i + 2 ];
              parsed.easing          = 'ease-in';
              break;
            }

            parsed.move   = words[ i + 1 ];
            parsed.easing = 'ease-in';
            break;

          case 'ease-in-out':

            if ( words[ i + 1 ] == 'up'
              || words[ i + 1 ] == 'down' ){

              parsed.scale.direction = words[ i + 1 ];
              parsed.scale.power     = words[ i + 2 ];
              parsed.easing          = 'ease-in-out';
              break;
            }

            parsed.move   = words[ i + 1 ];
            parsed.easing = 'ease-in-out';
            break;

          case 'ease-out':

            if ( words[ i + 1 ] == 'up'
              || words[ i + 1 ] == 'down' ){

              parsed.scale.direction = words[ i + 1 ];
              parsed.scale.power     = words[ i + 2 ];
              parsed.easing          = 'ease-out';
              break;
            }

            parsed.move   = words[ i + 1 ];
            parsed.easing = 'ease-out';
            break;

          case 'hustle':

            if ( words[ i + 1 ] == 'up'
              || words[ i + 1 ] == 'down' ){

              parsed.scale.direction = words[ i + 1 ];
              parsed.scale.power     = words[ i + 2 ];
              parsed.easing          = 'cubic-bezier( 0.6, 0.2, 0.1, 1 )';
              break;
            }

            parsed.move   = words[ i + 1 ];
            parsed.easing = 'cubic-bezier( 0.6, 0.2, 0.1, 1 )';
            break;

          case 'over':

            parsed.over = words[ i + 1 ];
            break;

          case 'flip':
          case 'pitch':
            parsed.rotate   = parsed.rotate || {};
            parsed.rotate.x = words[ i + 1 ];
            break;

          case 'spin':
          case 'yaw':
            parsed.rotate   = parsed.rotate || {};
            parsed.rotate.y = words[ i + 1 ];
            break;

          case 'roll':
            parsed.rotate   = parsed.rotate || {};
            parsed.rotate.z = words[ i + 1 ];
            break;

          case 'reset':

            if ( words[ i - 1 ] == 'no' ){
              parsed.reset = false;
            } else {
              parsed.reset = true;
            }
            break;

          case 'scale':

            parsed.scale = {};

            if ( words[ i + 1 ] == 'up'
              || words[ i + 1 ] == 'down' ){

              parsed.scale.direction = words[ i + 1 ];
              parsed.scale.power     = words[ i + 2 ];
              break;
            }

            parsed.scale.power = words[ i + 1 ];
            break;

          case 'vFactor':
          case 'vF':
            parsed.vFactor = words[ i + 1 ];
            break;

          case 'opacity':
            parsed.opacity = words[ i + 1 ];
            break;

          default:
            return;
        }
      });

      config = _extendClone( context, parsed );
    }

    // Now let’s prepare the configuration for CSS

    if ( config.easing == 'hustle' ){
      config.easing = 'cubic-bezier( 0.6, 0.2, 0.1, 1 )';
    }

    if ( config.enter === 'top'
      || config.enter === 'bottom' ){

      config.axis = 'Y';

    } else {
      config.axis = 'X';
    }

    // Let’s make sure our our pixel distances are negative for top and left.
    // e.g. "enter top and move 25px" starts at 'top: -25px' in CSS

    if ( config.enter === 'top'
      || config.enter === 'left' ){

      config.move = '-' + config.move;
    }

    return config;
  };



  ScrollReveal.prototype.styleFactory = function( elem ){

    var initial
      , inline
      , config
      , duration
      , original
      , reset
      , target
      , transition;

    // You can customize ScrollReveal with mobile-only logic here, for example,
    // change the starting opacity, or remove animation delay

    if ( self.isMobile() && self.defaults.mobile ){
      // Stuff...
    }

    // Capture the original inline styles.

    if ( !elem.styles ){

      original = elem.domEl.getAttribute('style')

    } else {

      original = elem.styles.original;
    }

    // Since JavaScript runs after the page has begun rendering, it’s possible
    // that elements will be seen before ScrollReveal can hide them during
    // initialization

    // One technique is to apply `visibility: hidden` via CSS to your elements
    // so they load hidden, allowing ScrollReveal to overwrite visibility here

    if ( original ){

      inline = original + '; visibility: visible; ';

    } else {

      inline = 'visibility: visible; ';
    }


    // Calculate animation duration (in milliseconds)

    config   = elem.config;
    duration = {};

    duration.reveal = ( parseFloat( config.over ) + parseFloat( config.wait ) ) * 1000
    duration.reset  =   parseFloat( config.over ) * 1000;

    // Build unprefixed and webkit transition styles

    transition = '-webkit-transition: -webkit-transform ' + config.over + ' ' + config.easing + ' ' + config.wait + ', opacity ' + config.over + ' ' + config.easing + ' ' + config.wait + '; ' +
                         'transition: transform '         + config.over + ' ' + config.easing + ' ' + config.wait + ', opacity ' + config.over + ' ' + config.easing + ' ' + config.wait + '; ' +
                '-webkit-perspective: 1000;' +
        '-webkit-backface-visibility: hidden;';

    // Build alternate transition styles with no delay, for animation resets

    reset      = '-webkit-transition: -webkit-transform ' + config.over + ' ' + config.easing + ' 0s, opacity ' + config.over + ' ' + config.easing + ' 0s; ' +
                         'transition: transform '         + config.over + ' ' + config.easing + ' 0s, opacity ' + config.over + ' ' + config.easing + ' 0s; ' +
                '-webkit-perspective: 1000; ' +
        '-webkit-backface-visibility: hidden; ';

    // Create initial and target animation styles

    initial = 'transform:';
    target  = 'transform:';

    generateStyles();

    // Create initial and target animation styles again, for webkit browsers

    initial += '-webkit-transform:';
    target  += '-webkit-transform:';

    generateStyles();

    return {
      initial:    initial,
      inline:     inline,
      duration:   duration,
      original:   original,
      reset:      reset,
      transition: transition,
      target:     target
    };

    function generateStyles(){

      if ( parseInt( config.move ) !== 0 ){
        initial += ' translate' + config.axis + '(' + config.move + ')';
        target  += ' translate' + config.axis + '(0)';
      }

      if ( parseInt( config.scale.power ) !== 0 ){

        if ( config.scale.direction === 'up' ){
          config.scale.value = 1 - ( parseFloat( config.scale.power ) * 0.01 );
        }

        else if ( config.scale.direction === 'down' ){
          config.scale.value = 1 + ( parseFloat( config.scale.power ) * 0.01 );
        }

        initial += ' scale(' + config.scale.value + ')';
        target  += ' scale(1)';
      }

      if ( config.rotate.x ){
        initial += ' rotateX(' + config.rotate.x + ')';
        target  += ' rotateX(0)';
      }

      if ( config.rotate.y ){
        initial += ' rotateY(' + config.rotate.y + ')';
        target  += ' rotateY(0)';
      }

      if ( config.rotate.z ){
        initial += ' rotateZ(' + config.rotate.z + ')';
        target  += ' rotateZ(0)';
      }

      initial += '; opacity: ' + config.opacity + '; ';
      target  += '; opacity: 1; ';
    }
  };



  ScrollReveal.prototype.getViewportHeight = function( viewport ){

    var client = viewport['clientHeight'];
    var inner  = window['innerHeight'];

    if ( viewport === window.document.documentElement ){
      return ( client < inner ) ? inner : client;
    }

    return client;
  };



  ScrollReveal.prototype.getScrolled = function( viewport ){

    if ( viewport === window.document.documentElement ){

      return window.pageYOffset;

    } else {

      return viewport.scrollTop + viewport.offsetTop;
    }
  };



  ScrollReveal.prototype.getOffset = function( domEl ){

    var offsetTop  = 0;
    var offsetLeft = 0;

    do {

      if ( !isNaN( domEl.offsetTop ) ){
        offsetTop  += domEl.offsetTop;
      }

      if ( !isNaN( domEl.offsetLeft ) ){
        offsetLeft += domEl.offsetLeft;
      }

    } while ( domEl = domEl.offsetParent );

    return {
      top: offsetTop,
      left: offsetLeft
    };
  };



  ScrollReveal.prototype.isElemVisible = function( elem ){

    var elHeight = elem.domEl.offsetHeight
      , elTop    = self.getOffset( elem.domEl ).top
      , elBottom = elTop + elHeight
      , vFactor  = elem.config.vFactor
      , viewport = elem.config.viewport;

    return ( confirmBounds() || isPositionFixed() );

    function confirmBounds(){

      var top        = elTop + elHeight * vFactor
        , bottom     = elBottom - elHeight * vFactor
        , viewTop    = self.getScrolled( viewport )
        , viewBottom = viewTop + self.getViewportHeight( viewport );

      return ( top < viewBottom ) && ( bottom > viewTop );
    }

    function isPositionFixed(){
      var style = elem.domEl.currentStyle || window.getComputedStyle( elem.domEl, null );
      return style.position === 'fixed';
    }
  };



  ScrollReveal.prototype.isMobile = function(){

    var agent = navigator.userAgent || navigator.vendor || window.opera;

    return (/(ipad|playbook|silk|android|bb\d+|meego).+mobile|avantgo|bada\/|blackberry|blazer|compal|elaine|fennec|hiptop|iemobile|ip(hone|od)|iris|kindle|lge |maemo|midp|mmp|mobile.+firefox|netfront|opera m(ob|in)i|palm( os)?|phone|p(ixi|re)\/|plucker|pocket|psp|series(4|6)0|symbian|treo|up\.(browser|link)|vodafone|wap|windows (ce|phone)|xda|xiino/i.test( agent )||/1207|6310|6590|3gso|4thp|50[1-6]i|770s|802s|a wa|abac|ac(er|oo|s\-)|ai(ko|rn)|al(av|ca|co)|amoi|an(ex|ny|yw)|aptu|ar(ch|go)|as(te|us)|attw|au(di|\-m|r |s )|avan|be(ck|ll|nq)|bi(lb|rd)|bl(ac|az)|br(e|v)w|bumb|bw\-(n|u)|c55\/|capi|ccwa|cdm\-|cell|chtm|cldc|cmd\-|co(mp|nd)|craw|da(it|ll|ng)|dbte|dc\-s|devi|dica|dmob|do(c|p)o|ds(12|\-d)|el(49|ai)|em(l2|ul)|er(ic|k0)|esl8|ez([4-7]0|os|wa|ze)|fetc|fly(\-|_)|g1 u|g560|gene|gf\-5|g\-mo|go(\.w|od)|gr(ad|un)|haie|hcit|hd\-(m|p|t)|hei\-|hi(pt|ta)|hp( i|ip)|hs\-c|ht(c(\-| |_|a|g|p|s|t)|tp)|hu(aw|tc)|i\-(20|go|ma)|i230|iac( |\-|\/)|ibro|idea|ig01|ikom|im1k|inno|ipaq|iris|ja(t|v)a|jbro|jemu|jigs|kddi|keji|kgt( |\/)|klon|kpt |kwc\-|kyo(c|k)|le(no|xi)|lg( g|\/(k|l|u)|50|54|\-[a-w])|libw|lynx|m1\-w|m3ga|m50\/|ma(te|ui|xo)|mc(01|21|ca)|m\-cr|me(rc|ri)|mi(o8|oa|ts)|mmef|mo(01|02|bi|de|do|t(\-| |o|v)|zz)|mt(50|p1|v )|mwbp|mywa|n10[0-2]|n20[2-3]|n30(0|2)|n50(0|2|5)|n7(0(0|1)|10)|ne((c|m)\-|on|tf|wf|wg|wt)|nok(6|i)|nzph|o2im|op(ti|wv)|oran|owg1|p800|pan(a|d|t)|pdxg|pg(13|\-([1-8]|c))|phil|pire|pl(ay|uc)|pn\-2|po(ck|rt|se)|prox|psio|pt\-g|qa\-a|qc(07|12|21|32|60|\-[2-7]|i\-)|qtek|r380|r600|raks|rim9|ro(ve|zo)|s55\/|sa(ge|ma|mm|ms|ny|va)|sc(01|h\-|oo|p\-)|sdk\/|se(c(\-|0|1)|47|mc|nd|ri)|sgh\-|shar|sie(\-|m)|sk\-0|sl(45|id)|sm(al|ar|b3|it|t5)|so(ft|ny)|sp(01|h\-|v\-|v )|sy(01|mb)|t2(18|50)|t6(00|10|18)|ta(gt|lk)|tcl\-|tdg\-|tel(i|m)|tim\-|t\-mo|to(pl|sh)|ts(70|m\-|m3|m5)|tx\-9|up(\.b|g1|si)|utst|v400|v750|veri|vi(rg|te)|vk(40|5[0-3]|\-v)|vm40|voda|vulc|vx(52|53|60|61|70|80|81|83|85|98)|w3c(\-| )|webc|whit|wi(g |nc|nw)|wmlb|wonu|x700|yas\-|your|zeto|zte\-/i.test( agent.substr( 0, 4 ))) ? true : false;
  };



  ScrollReveal.prototype.isSupported = function(){

    var sensor    = document.createElement('sensor')
      , cssPrefix = 'Webkit,Moz,O,'.split(',')
      , tests     = ( 'transition ' + cssPrefix.join('transition,') ).split(',');

    for ( var i = 0; i < tests.length; i++ ){
      if ( !sensor.style[ tests[ i ] ] === '' ){
        return false;
      }
    }

    return true;
  };



  ScrollReveal.prototype.cleanDOM = function(){
    for ( var key in self.store.elements ){
      if ( self.store.elements.hasOwnProperty( key ) ){
        self.store.elements[ key ].domEl.removeAttribute('data-sr data-sr-id');
      }
    }
  };

// Private Methods /////////////////////////////////////////////////////////////

  _updateElemStore = function( elem ){

    self.store.elements[ elem.id ] = elem;

    if ( self.store.viewports.indexOf( elem.config.viewport ) == -1 ){
      self.store.viewports.push( elem.config.viewport );
    }
  };

  _handler = function( e ){

    if ( !self.blocked ){

      self.blocked = true;
      _requestAnimFrame( self.animate );
    }
  };

  _extend = function( target, src ){

    for ( var prop in src ){
      if ( src.hasOwnProperty( prop ) ){
        target[ prop ] = src[ prop ];
      }
    }

    return target;
  };

  _extendClone = function( target, src ){

    var clone = {};

    for ( var prop in target ){
      if ( target.hasOwnProperty( prop ) ){
        clone[ prop ] = target[ prop ];
      }
    }

    for ( var prop in src ){
      if ( src.hasOwnProperty( prop ) ){
        clone[ prop ] = src[ prop ];
      }
    }

    return clone;
  };

  _isDOMElement = function( obj ){
    return (
      typeof HTMLElement === "object" ? obj instanceof HTMLElement : // DOM2
      obj && typeof obj === "object" && obj !== null && obj.nodeType === 1 && typeof obj.nodeName==="string"
    );
  };

  // Polyfill //////////////////////////////////////////////////////////////////

  _requestAnimFrame = (function(){

    return window.requestAnimationFrame        ||
           window.webkitRequestAnimationFrame  ||
           window.mozRequestAnimationFrame     ||

          function( callback ){
            window.setTimeout( callback, 1000 / 60 );
          };
  }());

  //////////////////////////////////////////////////////////////////////////////

  return ScrollReveal;

})( window );
